jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,224 @@
+
+`newChannel` is a generic proc - you can specify the types of the data you want to send or receive.
+
+```nim
+var chan1 = newChannel[int]()
+# or
+var chan2 = newChannel[string](elements = 1) # unbuffered channel
+# or
+var chan3 = newChannel[seq[string]](elements = 30) # buffered channel
+```
+
+The `send` proc takes data that we want to send to the channel. The passed data is moved around, not copied. Because `chan.send(isolate(data))` is very common to use, `template send[T](c: var Chan[T]; src: T) = chan.send(isolate(src))` is provided for convenience. For example, you can use `chan.send("Hello World")` instead of `chan.send(isolate("Hello World!"))`.
+
+There are two useful procs for a receiver: `recv` and `tryRecv`. `recv` blocks until something is sent to the channel. In contrast, `tryRecv` doesn't block - if no message exists in the channel, it just fails and returns `false`. We can write a while loop to call `tryRecv`and handle a message when available.
+
+### It is safe and convenient
+
+The Nim compiler rejects the program below at compile time. It says that `expression cannot be isolated: s`. This is because `s` is a `ref object` - it may be modified somewhere and is not unique, so the variable cannot be isolated.
+
+
+```nim
+import std/[channels, json, isolation]
+
+var chan = newChannel[JsonNode]()
+
+proc spawnCrawlers =
+  var s = newJString("Hello, Nim")
+  chan.send isolate(s)
+```
+
+It is only allowed to pass a function call directly.
+
+```nim
+import std/[channels, json, isolation]
+
+var chan = newChannel[JsonNode]()
+
+proc spawnCrawlers =
+  chan.send isolate(newJString("Hello, Nim"))
+```
+
+`Isolated` data can only be moved, not copied. It is implemented as a library without bloating Nim's core type system. The `isolate` proc is used to create an isolated subgraph from the expression `value`. Whether the expression `value` is isolated is checked at compile time. The `extract` proc is used to get the internal value of `Isolated` data.
+
+```nim
+import std/isolation
+
+var data = isolate("string")
+doAssert data.extract == "string"
+doAssert data.extract == ""
+```
+
+By means of `Isolated` data, the channels become safer and more convenient to use.
+
+
+## Benchmark
+
+Here is a simple benchmark. We create 10 threads that send data to the channel and 5 threads that receive it.
+
+```nim
+# benchmark the new channel implementation with 
+# `nim c -r --threads:on --gc:orc -d:newChan -d:danger app.nim`
+#
+# benchmark the old channel implementation with
+# `nim c -r --threads:on -d:oldChan -d:danger app.nim`
+
+import std/times
+
+var
+  sender: array[0 .. 9, Thread[void]]
+  receiver: array[0 .. 4, Thread[void]] 
+when defined(newChan):
+  import std/[channels, isolation]
+  var chan = newChannel[seq[string]](40)
+
+  proc sendHandler() =
+    chan.send(isolate(@["Hello, Nim"]))
```

Comment:
```suggestion
    chan.send(@["Hello, Nim"])
```
because `send` template does this for you.
Then you can move `sendHandler` outside of the contional `when defined(newChan):` since that code becomes identical.

Simpler = better.

Reply:
fixed